draft-ietf-quic-recovery.md

```diff
@@ -763,24 +763,33 @@ The RECOMMENDED value is 2 * max_datagram_size.
 
 ## Slow Start
 
-While in slow start, QUIC increases the congestion window by the
-number of bytes acknowledged when each acknowledgment is processed, resulting
-in exponential growth of the congestion window.
+The congestion controller is in slow start any time the congestion window is
+below the slow start threshold.  The congestion controller begins in slow start.
 
-QUIC exits slow start upon loss or upon increase in the ECN-CE counter.
-When slow start is exited, the congestion window halves and the slow start
-threshold is set to the new congestion window.  QUIC re-enters slow start
-any time the congestion window is less than the slow start threshold,
-which only occurs after persistent congestion is declared.
+While the congestion controller is in slow start, the congestion window
+increases by the number of bytes acknowledged when each acknowledgment is
+processed.  This results in exponential growth of the congestion window.
+
+The congestion controller exits slow start and enters congestion avoidance when
+a packet is lost or when the ECN-CE count increases.
+
+The congestion controller re-enters slow start any time the congestion window is
+less than the slow start threshold, which only occurs after persistent
+congestion is declared.
 
 ## Congestion Avoidance
 
-Slow start exits to congestion avoidance.  Congestion avoidance uses an
-Additive Increase Multiplicative Decrease (AIMD) approach that increases
-the congestion window by one maximum packet size per congestion window
-acknowledged.  When a loss or ECN-CE marking is detected, NewReno halves
-the congestion window, sets the slow start threshold to the new
-congestion window, and then enters the recovery period.
+The congestion controller is in congestion avoidance any time the congestion
+window is at or above the slow start threshold.
+
+Congestion avoidance uses an Additive Increase Multiplicative Decrease (AIMD)
+approach that increases the congestion window by one maximum packet size per
+congestion window acknowledged.
+
+When first entering congestion avoidance, when a packet is declared lost, or
+when the ECN-CE count is detected, the congestion controller halves the
+congestion window, reduces the slow start threshold to the resulting size, and
+enters the recovery period.
```

Comment:
I'm just realising this now but this is actually not how TCP works. When you enter recovery you halve the ssthresh (basically to remember the right value for later). Then what TCP does is that is keeps sending one new packet for each ACK received (conservation principle: not know which or how many packets have been ACKed but know there must be at least one successful received and leaving the network) and then when all losses are repaired it sets the cwnd to ssthresh and leaves recovery. If you halve cwnd immediately when entering recovery, you can't send data for half an RTT until you received enough ACKs to be below the cwnd. This is usually not what you want.

Reply:
I think you're correct, do you have a good RFC reference handy?  I couldn't easily find this text.

Reply:
https://tools.ietf.org/html/rfc5681#section-4.3

Reply:
@janaiyengar just explained to me why this difference exists, and I found the argument convincing.  I don't know whether this is worth calling out though, it seems like it might be a little tricky to get down in text.

Reply:
I think the MUST requirement is that the cwnd is set to ssthresh when LEAVING recovery. In order to decrement the cwnd gradually we could say that only one new packet for two ACKed ones should be send which then would be mostly inline with TCP or we can leave the detailed scheme to the implementation and just put some warnings to not overload the link in recovery. Alternatively you could also e.g. halve the cwnd and then pace out packets over one RTT (which is assume is what implementation do right now?)

Reply:
If we're going to make a change, I think it needs to be on #3978, which is a design PR that's going to be sent out for consensus call soon.

Reply:
Yes.  This PR is intended to be a strictly editorial conversion.  I hope that I have faithfully translated the text from #3978.

Reply:
(This is being discussed in #3978.)